dcos_test_utils/gce.py

```diff
@@ -0,0 +1,174 @@
+                                                       instance=instance_name, body=body)
+        response = request.execute()
+        return response
+
+
+class Instance:
+    def __init__(self, gce_wrapper, name, priv_ip, pub_ip):
+        self.gce_wrapper = gce_wrapper
+        self.name = name
+        self.priv_ip = priv_ip
+        self.pub_ip = pub_ip
+
+    def get_private_public_ip(self):
+        return {
+            'private_ip': self.priv_ip,
+            'public_ip': self.pub_ip}
+
+
+class Deployment:
+    def __init__(self, gce_wrapper, name, zone):
+        self.gce_wrapper = gce_wrapper
+        self.name = name
+        self.zone = zone
+        self.instances = []
+        self.fingerprint = None
+
+    @catch_http_exceptions
+    def get_instances(self):
+        if not self.instances:
+            instance_info_list = self.gce_wrapper.list_deployment_instances(self)
+            for i in instance_info_list:
+                self.instances.append(Instance(self.gce_wrapper, i['name'],
+                                               i['networkInterfaces'][0]['networkIP'],
+                                               i['networkInterfaces'][0]['accessConfigs'][0]['natIP']))
+        return self.instances
+
+    @catch_http_exceptions
+    def get_fingerprint(self):
+        if not self.fingerprint:
+            self.fingerprint = self.get_info()['fingerprint']
+        return self.fingerprint
+
+    @catch_http_exceptions
+    def get_host_ips(self):
+        return [i.get_private_public_ip() for i in self.instances]
+
+    @catch_http_exceptions
+    def apply_ssh_key(self, public_key):
+        body = {
+            "kind": "compute#metadata",
+            'fingerprint': self.get_fingerprint(),
+            'items': [
+                {
+                    'key': 'ssh-keys',
+                    'value': 'dcos:ssh-rsa ' + public_key + ' dcos'
+                }
+            ]
+        }
+
+        for i in self.get_instances():
+            yield self.gce_wrapper.set_metadata(self, body, i['name'])
+
+    @catch_http_exceptions
+    def get_info(self):
+        return self.gce_wrapper.get_deployment_info(self)
+
+    @catch_http_exceptions
+    def delete(self):
+        return self.gce_wrapper.delete_deployment(self)
+
+    def _check_status(response):
+        status = response['operation']['status']
+        if status == 'DONE':
+            return False
+        elif status == 'RUNNING' or status == 'PENDING':
+            print('Waiting for deployment')
+            return True
+        else:
+            raise Exception('Deployment failed with response: ' + str(response))
+
+    @catch_http_exceptions
+    @retry(wait_fixed=2000, retry_on_result=_check_status, stop_max_attempt_number=15)
+    def wait(self):
+        response = self.get_info()
+        errors = response['operation'].get('errors')
+        if errors:
+            self.delete()
```

Comment:
This auto-delete code is out of place here. This is quite high-level logic in a low level function. In general, I do not expect my poller's to mutate my resources.